File: dotNet/Lib/ContextBuilder.cs
using System.Collections.Generic;
using System.Linq;

namespace Com.Company.Context {
    public class ContextBuilder : IContextBuilder {
        public IContext Build(object config) {
            if (config == null) {
                throw new IllegalConfigException("Configuration is null");
            }

            var methodList = EnumerateBeanGetters(config);
            if (!methodList.Any()) {
                throw new IllegalConfigException("Configuration does not contains beans");
            }

            var beanList = new List<ConfigBeanItem>();
            foreach(var methodItem in methodList) {
                CreateInstance(config, methodItem, methodList, beanList);
            }

            return new Context(beanList);
        }

        private IList<MethodBeanItem> EnumerateBeanGetters(object config) {
            var result = new List<MethodBeanItem> ();

            var configType = config.GetType();
            var methodArray = configType.GetMethods();
            foreach (var methodItem in methodArray) {
                if (methodItem.Name.StartsWith("Get")) {
                    var attrs = methodItem.GetCustomAttributes(typeof(Bean), false);
                    if (attrs != null && attrs.Any()) {
                        if (methodItem.ReturnType == typeof(void)) {
                            var msg = string.Format(
                                "Bean {0} has void return type",
                                methodItem.Name);

                            throw new IllegalBeanException(msg);
                        }

                        var beanAttr = attrs[0] as Bean;
                        if (result.Any(x => x.Id.Equals(beanAttr.Id) && x.Method.ReturnType.Equals(methodItem.ReturnType))) {
                            string msg = string.Format(
                                "Bean are duplicated. Id: '{0}', type: {1}",
                                beanAttr.Id,
                                methodItem.ReturnType);

                        throw new IllegalConfigException(msg);                        }

                        result.Add(new MethodBeanItem(methodItem, beanAttr.Id));
                    }
                }
            }

            return result;
        }

        private object CreateInstance(object config,
                                      MethodBeanItem methodItem,
                                      IList<MethodBeanItem> methodList,
                                      IList<ConfigBeanItem> beanList) {

            object instance;
            if (!methodItem.Method.GetParameters().Any()) {
                instance = CreateWithoutParameters(config, methodItem, beanList);
            } else {
                instance = CreateWithParameters(config, methodItem, methodList, beanList);
            }

            beanList.Add(new ConfigBeanItem(methodItem.Id, methodItem.Method.ReturnType, instance));

            return instance;
        }

        private object CreateWithoutParameters(object config,
                                               MethodBeanItem methodItem,
                                               IList<ConfigBeanItem> beanList) {

            var beanItem = beanList.FirstOrDefault(x => x.Id.Equals(methodItem.Id) &&
                x.Interface.Equals(methodItem.Method.ReturnType));

            object instance;
            if (beanItem == null) {
                instance = methodItem.Method.Invoke(config, new object[] {});
            } else {
                instance = beanItem.Instance;
            }

            return instance;
        }

        private object CreateWithParameters(object config,
                                            MethodBeanItem methodItem,
                                            IList<MethodBeanItem> methodList,
                                            IList<ConfigBeanItem> beanList) {

            var paramsArray = methodItem.Method.GetParameters();
            object[] prms = new object[paramsArray.Length];
            var index = 0;
            foreach(var paramItem in paramsArray) {
                var paramId = string.Empty;
                var paramAnnotation = paramItem.GetCustomAttributes(typeof(BeanParameter), false);
                if (paramAnnotation != null && paramAnnotation.Any()) {
                    var annotation = paramAnnotation[0] as BeanParameter;
                    paramId = annotation.Id;
                }

                object instance;
                var paramObjectItem = beanList.FirstOrDefault(x => x.Id.Equals(paramId) &&
                    x.Interface.Equals(paramItem.ParameterType));

                if (paramObjectItem == null) {
                    var paramMethodItem = methodList.FirstOrDefault(x => x.Id.Equals(paramId) &&
                        x.Method.ReturnType.Equals(paramItem.ParameterType));
                    if (paramMethodItem == null) {
                        var msg = string.Format("Could not find the Bean for {0} of type {1}",
                                                methodItem.Method.Name,
                                                paramItem);

                        throw new IllegalBeanException(msg);
                    }

                    instance = CreateInstance(config, paramMethodItem, methodList, beanList);
                } else {
                    instance = paramObjectItem.Instance;
                }

                prms[index++] = instance;
            }

            return methodItem.Method.Invoke(config, prms);
        }
    }
}

File: java/src/com/company/context/ContextBuilder.java
package com.company.context;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Parameter;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class ContextBuilder implements IContextBuilder {
    public IContext build(Object config) throws
        IllegalBeanException,
        IllegalConfigException,
        IllegalAccessException,
        IllegalArgumentException,
        InvocationTargetException {

        if (config == null) {
            throw new IllegalConfigException("Configuration is null");
        }

        List<MethodBeanItem> methodList = enumerateBeanGetters(config);
        if (methodList.isEmpty()) {
            throw new IllegalConfigException("Configuration is empty");
        }

        List<ConfigBeanItem> beanList = new ArrayList<>();
        for (MethodBeanItem methodItem : methodList) {
            createInstance(config, methodItem, methodList, beanList);
        }

        return new Context(beanList);
    }

    private List<MethodBeanItem> enumerateBeanGetters(Object config) throws
        IllegalConfigException {

        List<MethodBeanItem> result = new ArrayList<>();

        Class<?> clazz = config.getClass();
        Method[] methodArray = clazz.getMethods();
        for (Method methodItem : methodArray) {
            if (methodItem.getName().startsWith("get")) {
                Bean annotation = methodItem.getAnnotation(Bean.class);
                if (annotation != null) {
                    if (result.stream().anyMatch((item) -> {
                        return item.mId.equals(annotation.id())
                                && item.mMethod.getReturnType().equals(methodItem.getReturnType());
                    })) {
                        String msg = String.format(
                                "Bean are duplicated. Id: '%s', type: %s",
                                annotation.id(),
                                methodItem.getReturnType());

                        throw new IllegalConfigException(msg);
                    }

                    if (methodItem.getReturnType().equals(void.class)) {
                        String msg = String.format("The bean %s has void type",
                                                    methodItem.getName());

                        throw new IllegalConfigException(msg);
                    }

                    result.add(new MethodBeanItem(methodItem, annotation.id()));
                }
            }
        }

        return result;
    }

    private Object createWithoutParameters(Object config,
                                           MethodBeanItem methodItem,
                                           List<ConfigBeanItem> beanList) throws
        IllegalAccessException,
        IllegalArgumentException,
        InvocationTargetException {

        Object instance;
        Optional<ConfigBeanItem> optional = beanList.stream().filter(
                (item) ->
                        item.mId.equals(methodItem.mId) &&
                        item.mInterface.equals(methodItem.mMethod.getReturnType())).findFirst();

        if (!optional.isPresent()) {
            instance = methodItem.mMethod.invoke(config);
        } else {
            instance = optional.get();
        }

        return instance;
    }

    private Object createWithParameters(Object config,
                                        MethodBeanItem methodItem,
                                        List<MethodBeanItem> methodList,
                                        List<ConfigBeanItem> beanList) throws
            IllegalBeanException,
            InvocationTargetException,
            IllegalAccessException {

        Parameter[] paramArray = methodItem.mMethod.getParameters();
        Object[] params = new Object[paramArray.length];
        int index = 0;
        for (Parameter paramItem : paramArray) {
            String paramId = "";
            BeanParameter paramAnnotation = paramItem.getAnnotation(BeanParameter.class);
            if (paramAnnotation != null) {
                paramId = paramAnnotation.id();
            }

            Object paramObject = findInstance(paramId, paramItem.getType(), beanList);
            if (paramObject == null) {
                MethodBeanItem paramMethodItem = findMethodItem(paramId,
                        paramItem.getType(),
                        methodList);
                if (paramMethodItem == null) {
                    String msg = String.format("Could not find the Bean for %s of type %s",
                            methodItem.mMethod.getName(),
                            paramItem);

                    throw new IllegalBeanException(msg);
                }

                paramObject = createInstance(config, paramMethodItem, methodList, beanList);
            }

            params[index++] = paramObject;
        }

        return methodItem.mMethod.invoke(config, params);
    }

    private Object createInstance(Object config,
                                  MethodBeanItem methodItem,
                                  List<MethodBeanItem> methodList,
                                  List<ConfigBeanItem> beanList) throws
        IllegalAccessException,
        IllegalArgumentException,
        InvocationTargetException,
        IllegalBeanException {

        Object instance;
        if (methodItem.mMethod.getParameterTypes().length == 0) {
            instance = createWithoutParameters(config, methodItem, beanList);
        } else {
            instance = createWithParameters(config, methodItem, methodList, beanList);
        }

        beanList.add(new ConfigBeanItem(methodItem.mId,
                                        methodItem.mMethod.getReturnType(),
                                        instance));

        return instance;
    }

    private Object findInstance(String id, Class<?> clazz, List<ConfigBeanItem> beanList) {
        Optional<ConfigBeanItem> result = beanList.stream().filter(
            (item) -> item.mId.equals(id) && item.mInterface.equals(clazz)).findFirst();

        return result.map(configBeanItem -> configBeanItem.mInstance).orElse(null);
    }

    private MethodBeanItem findMethodItem(String id,
                                          Class<?> clazz,
                                          List<MethodBeanItem> methodList) {
        Optional<MethodBeanItem> result = methodList.stream().filter(
            (item) -> item.mId.equals(id) && item.mMethod.getReturnType().equals(clazz))
            .findFirst();

        return result.orElse(null);
    }
}